drivers/cache/cache_andes.c

```diff
@@ -0,0 +1,737 @@
+/*
+ * Copyright (c) 2024 Andes Technology Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "soc_v5.h"
+
+#include <zephyr/init.h>
+#include <zephyr/kernel.h>
+#include <zephyr/arch/cpu.h>
+#include <zephyr/drivers/syscon.h>
+#include <zephyr/drivers/cache.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/arch/riscv/csr.h>
+
+LOG_MODULE_REGISTER(cache_andes, CONFIG_CACHE_LOG_LEVEL);
+
+/* L1 CCTL Command */
+#define CCTL_L1D_VA_INVAL	0
+#define CCTL_L1D_VA_WB		1
+#define CCTL_L1D_VA_WBINVAL	2
+#define CCTL_L1D_WBINVAL_ALL	6
+#define CCTL_L1D_WB_ALL		7
+#define CCTL_L1I_VA_INVAL	8
+#define CCTL_L1D_INVAL_ALL	23
+#define CCTL_L1I_IX_INVAL	24
+
+#define L2C_EXISTS	DT_NODE_HAS_COMPAT_STATUS(DT_NODE_EXISTS(	\
+			DT_NODELABEL(l2_cache)), andestech_l2c, okay)
```

Comment:
this means the node must be named as `l2_cache` in the devicetree

Reply:
Thanks for the suggestions!

1. I'll add the binding file, and check L2 cache DTS node by compatible
2. I'm not sure about the issues in `DT_INST_FOREACH_STATUS_OKAY` and `SYS_INIT`. 
    There are two macros: `SYS_INIT` and `DEVICE_DT_INST_DEFINE` for setting drivers up for boot time initialization. 
    `DEVICE_DT_INST_DEFINE` which will create a driver object isn't needed for the cache driver.

    `DT_INST_FOREACH_STATUS_OKAY` is used in the case that there are multiple nodes. However, in the current 
     implementation of the Andes cache driver, the focus is solely on a single L2 cache node.

Reply:
> 2\. `DT_INST_FOREACH_STATUS_OKAY` is used in the case that there are multiple nodes.

Sorry that was a typo - should be `DEVICE_DT_INST_DEFINE` / `DEVICE_DT_DEFINE` instead

> However, in the current implementation of the Andes cache driver, the focus is solely on a single L2 cache node.

Yeah they both work as fine for a single instance device driver like this without a device API, it's just that the `DT_` variant is designed for device drivers and opens to more possibilities in the future. It's nice to refactor this driver a bit at the very beginning to make use of the device driver macros and make these

```c
static struct cache_config cache_cfg; /* this can go to the 'config' of the device */
static struct k_spinlock lock; /* this can go to the 'data' of the device */
```

instance based so that the driver can be made multi-instance very easily in the future.

But these are all just suggestions.

Reply:
We have explored the concept of a multi-instance-based driver, but its viability is challenged when confronted with the introduction of a multi-level cache system. Simply parsing the L2 cache node is insufficient to address this complexity.

In a multi-level cache setup, there could be a scenario where each core has its private L2 cache, or two cores share an L2 cache, or with an additional L3 cache. This introduces a set of challenges: determining which level to treat as instances and, when performing cache operations, identifying the appropriate next-level cache. If we limit our focus to parsing the L2 cache exclusively, we not only overlook scenarios involving multiple L2 cache nodes but also make the code less adaptable for future refactoring.